### PR TITLE
Use testcontainers

### PR DESCRIPTION
**Pull request by bcarter97**
## Related Issues

- Connects to #

## Definition of Done:

The Below tasks should be completed before marking a PR as ready for review.

- [ ] Added / Removed / Updated relevant tests
- [ ] Scaladoc updated

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -31,13 +30,13 @@ lazy val root = (project in file("."))
       Cats.core,
       Cats.effect,
       Cats.log4cats,
-      Cats.log4catsSlf4j,
+      Cats.log4catsSlf4j  % Test,
```
Comment (bcarter97) [inline build.sbt:33]: This is only used for testing to actually print Logs. In the real world the user will provide the logging backend, we shouldn't assume Slf4j.

File: src/test/scala/integration/TopicLoaderIntSpec.scala
Comment (bcarter97) [inline src/test/scala/integration/TopicLoaderIntSpec.scala:1]: 90% of this file is just reformatting after changing the withKafkaContext line

File: build.sbt [hunk 2]
```diff
@@ -31,13 +30,13 @@ lazy val root = (project in file("."))
...
       Fs2.core,
       Fs2.kafka,
-      embeddedKafka,
-      scalaTest,
-      catsEffectTesting,
-      logbackClassic
+      catsEffectTesting   % Test,
+      logbackClassic      % Test,
```
Comment (bcarter97) [inline build.sbt:37]: Same as above, we shouldn't assume the logging backend, and this is only needed for tests

File: src/test/scala/utils/KafkaTestContainer.scala [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+import cats.effect.syntax.all.*
+import cats.effect.{Ref, Resource, Sync}
+import cats.syntax.all.*
+import org.scalatest.Assertion
+import org.testcontainers.containers.KafkaContainer as Underlying
+import org.testcontainers.utility.DockerImageName
+import utils.KafkaContainer.KafkaConfig
+
+trait KafkaTestContainer[F[_]] {
+  def withKafkaContext(test: KafkaConfig => F[Assertion])(using Sync[F]): F[Assertion] =
```
Comment (lacarvalho91) [inline src/test/scala/utils/KafkaTestContainer.scala:12]: i dont find this name very clear, i think `withRunningKafka` is clearer

**Review by lacarvalho91 — APPROVED** [review]
one tiny comment but LGTM

File: src/test/scala/utils/KafkaTestContainer.scala (hunk 3 above)
Comment (bcarter97) [reply src/test/scala/utils/KafkaTestContainer.scala]: https://github.com/sky-uk/fs2-kafka-topic-loader/pull/105/commits/d5f8271fff8442b76fe9f4350f2b3aabf7512450

File: it/src/test/scala/integration/LoadExampleIntSpec.scala [hunk 4]
```diff
@@ -4,43 +4,48 @@ import base.KafkaSpecBase
...
+          _      <- publishStringMessage(inputTopic, "key2", "value2")
+          _      <- runAppAndDiscard
+          result <- consumeStringMessage(outputTopic, autoCommit = true)
+        } yield result shouldBe "value2"
+      }
     }
   }
 
-  private trait TestContext {
+  final case class TestContext()(using kafkaConfig: KafkaConfig) {
```
Comment (MarkTickner) [inline it/src/test/scala/integration/LoadExampleIntSpec.scala:48]:
```suggestion
  final case class TestContext()(using KafkaConfig) {
```
as the below consumerSettings and producerSettings where it is used, names it for internal scope

File: src/test/scala/utils/KafkaTestContainer.scala [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+  sealed trait Running[F[_]] {
+    def config: KafkaConfig
+    def stop: F[Unit]
+  }
+
+  def apply[F[_]](using F: Sync[F]): Resource[F, Running[F]] =
+    Resource.make {
+      for {
+        status           <- Ref.of(Status.Starting)
+        container        <- Either.catchNonFatal(Underlying(DockerImageName.parse("confluentinc/cp-kafka:7.4.0"))).liftTo[F]
```
Comment (MarkTickner) [inline src/test/scala/utils/KafkaTestContainer.scala:26]: Is it worth having the docker image name/version in config or something? Seems a bit weird having it here directly

Comment (MarkTickner) [reply src/test/scala/utils/KafkaTestContainer.scala:26]: I've also not used testcontainers before so not sure of conventions etc

**Pull request merged**